### PR TITLE
Update available-rules.md

### PR DESCRIPTION
**Pull request by vtchem**
## Which rule and warnings did you add?

This pull request closes 

## Actions checklist
* [ ] Implemented Rule, added Warnings
* [ ] Added tests on checks
* [ ] Added tests on fixers
* [ ] Updated diktat-analysis.yml
* [ ] Updated available-rules.md

### CODE REVIEW

**Review by nulls — APPROVED** [review]

File: info/available-rules.md [hunk 1]
```diff
@@ -1,131 +1,131 @@
...
+| 1    | 1.1.1    | FILE_NAME_INCORRECT                       | Check: warns if a file name does not have an extension .kt/.kts<br>Fix: no                                                                                                                                                                                        | no  | no                                                                                                                                                                                                                         | Extensions should be configurable; it can be autofixed                                                                                                                                                             |
```
Comment (0x6675636b796f75676974687562) [inline info/available-rules.md:8]: Check: warns if a file name does not have a `.kt`/`.kts` extension.

Fix: **none**.

File: info/available-rules.md [hunk 2]
```diff
@@ -1,131 +1,131 @@
...
+| 1    | 1.1.1    | GENERIC_NAME                              | Check: warns if a generic name contains more than 1 letter (capital). It can be followed by numbers, for example: T12, T<br>Fix:                                                                                                                                   | yes | no                                                                                                                                                                                                                         | Recursively update usages of this identifier in the project                                                                                                                                                                   |
```
Comment (0x6675636b796f75676974687562) [inline info/available-rules.md:9]: > Check: warns if **the** name of **a** [_generic type parameter_](https://kotlinlang.org/docs/generics.html) (e.g: `T`) consists of more than a single capital letter. The capital letter can be followed by numbers, though, e. g.: `T12`.

File: info/available-rules.md [hunk 3]
```diff
@@ -1,131 +1,131 @@
...
+| 1    | 1.1.1    | VARIABLE_HAS_PREFIX                       | Check: warns if a variable has a prefix (like mVariable or M_VARIABLE), generally it is a bad code style (Android - is the only exception<br>Fix: no fix is available to automatically create a name                                                      | no  | no                                                                                                                                                                                                                         | Maybe we can fix it, but I do not see any sense for it                                                                                                                                                                      |
```
Comment (0x6675636b796f75676974687562) [inline info/available-rules.md:10]: Check: warns if **the** name of **a** variable has a [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation) specific prefix (like `mVariable` or `M_VARIABLE`); generally, it is **considered** a bad code style (_Android_ **being** the the only exception).

Fix: none is available, as only a human can choose a meaningful name depending on the context.

File: info/available-rules.md [hunk 4]
```diff
@@ -1,131 +1,131 @@
...
+| 1    | 1.2.1    | PACKAGE_NAME_MISSING                      | Check: warns if the package name is missing in a file<br>Fix: automatically adds a package directive with the name that starts from the domain name (example - com.huawei) and contains the real directory                                                      | yes | no                                                                                                                                                                                                                         | Recursively fix all imports in the project.<br>Fix the directory where the code is stored.<br>Make this check isolated from the domain name addition.                                                                                  |
```
Comment (0x6675636b796f75676974687562) [inline info/available-rules.md:11]: > The check warns if a file is missing the package declaration.

**Review by 0x6675636b796f75676974687562 — CHANGES_REQUESTED** [review]

File: info/available-rules.md [hunk 5]
```diff
@@ -1,131 +1,131 @@
...
+| 1    | 1.1.1    | BACKTICKS_PROHIBITED                      | Check: warns if backticks (``) are used in the identifier name, except in the case when it is a test method (marked with @Test annotation)                                                                                                                        | no  | no                                                                                                                                                                                                                         | -                                                                                                                                                                                                                             | |
```
[the remark this answers is not in the document]
Comment (vtchem) [reply info/available-rules.md:4]: OK

File: info/available-rules.md [hunk 6]
```diff
@@ -1,131 +1,131 @@
...
+| 1    | 1.1.1    | EXCEPTION_SUFFIX                          | Check: warns if a class that extends any Exception class does not have an Exception suffix<br>Fix: Adding suffix "Exception" to a class name                                                                                                                      | yes | no                                                                                                                                                                                                                         | Need to add tests for this                                                                                                                                                                                                    |
```
[the remark this answers is not in the document]
Comment (vtchem) [reply info/available-rules.md:6]: OK

File: info/available-rules.md [hunk 7]
```diff
@@ -1,131 +1,131 @@
...
+| 1    | 1.1.1    | IDENTIFIER_LENGTH                         | Check: identifier length should be in range [2,64] except names that are used in industry like {i, j} and 'e' for catching exceptions.<br>Fix: no fix is available to automatically create a name                                                         | no  | no                                                                                                                                                                                                                         | May be make this rule configurable (length)                                                                                                                                                                                   |
```
[the remark this answers is not in the document]
Comment (vtchem) [reply info/available-rules.md:7]: OK

File: info/available-rules.md (hunk 1 above)
Comment (vtchem) [reply info/available-rules.md:8]: OK

File: info/available-rules.md (hunk 2 above)
Comment (vtchem) [reply info/available-rules.md:9]: Check: warns if the name of a  [_generic type parameter_](https://kotlinlang.org/docs/generics.html) (e.g. `T`) consists of more than a single capital letter. The capital letter can be followed by numbers, though (e. g. T12), T

File: info/available-rules.md (hunk 3 above)
Comment (vtchem) [reply info/available-rules.md:10]: Question: Do we need the following?
generally, it is **considered** a bad code style (_Android_ **being** the the only exception).

File: info/available-rules.md (hunk 4 above)
Comment (vtchem) [reply info/available-rules.md:11]: 1. OK
2. **Question**: What does the following sentence mean?
Maybe we can fix it, but I do not see any sense for it.

File: info/available-rules.md [hunk 8]
```diff
@@ -1,131 +1,131 @@
...
+| 1    | 1.1.1    | CONFUSING_IDENTIFIER_NAMING               | Check: warns if the identifier has inappropriate name (See table of rule 1.2 part 6).                                                                                                                                                                            | no  | no                                                                                                                                                                                                                         | no                                                                                                                                                                                                                            |
```
[the remark this answers is not in the document]
Comment (vtchem) [reply info/available-rules.md]: OK

File: info/available-rules.md [hunk 9]
```diff
@@ -1,131 +1,131 @@
...
+| 1    | 1.2.1    | PACKAGE_NAME_INCORRECT_SYMBOLS            | Check: warns if a package name has incorrect symbols like underscore or non-ASCII letters/digits.Exception: underscores that are used for differentiating of keywords in a name.<br>Fix: no fix currently; will be suggested later                              | no  | no                                                                                                                                                                                                                         | Add autofix for at least converting underscore to a dot or replacing itFix the directory where the code is stored.Cover autofix with tests                                                                                    |
```
Comment (vtchem) [inline info/available-rules.md:15]: "..replacing itFix the directory where the code is stored." 
What does it mean?

File: info/available-rules.md [hunk 10]
```diff
@@ -1,131 +1,131 @@
...
+| 1    | 1.2.1    | INCORRECT_PACKAGE_SEPARATOR               | Check: warns if the underscore is incorrectly used to split the package naming<br>Fix: fixing all nodes in AST and the package name to remove all underscores                                                                                                        | no  | no                                                                                                                                                                                                                         | Recursively update usages of this class in the project                                                                                                                                                                        |
```
Comment (vtchem) [inline info/available-rules.md:17]: "Check: warns if the underscore is incorrectly used to split the package naming." This is not clear to me. Please, explain.

File: info/available-rules.md [hunk 11]
```diff
@@ -1,131 +1,131 @@
...
+| 1    | 1.3.1    | CLASS_NAME_INCORRECT                      | Check: warns if the Class/Enum/Interface name does not match Pascal case ("([A-Z][a-z0-9]+)+")<br>Fix: fixing the case: if it is some fixed case (like snake/camel) - with word saving; if not - will restore PascalCase as is.                          | yes | no                                                                                                                                                                                                                         | Recursively update usages of this class in the projectCheck and decide the better way of converting the identifier to PascalCaseNeed to add checks using natural language processing to check that the class name contains only nouns |
```
Comment (vtchem) [inline info/available-rules.md:18]: Unclear sentence: "If it is some fixed case (like snake/camel) - with word saving; if not - will restore PascalCase as is." Please, explain.

File: info/available-rules.md [hunk 12]
```diff
@@ -1,131 +1,131 @@
...
+| 1    | 1.6.2    | FUNCTION_BOOLEAN_PREFIX                   | Check: functions/methods that return boolean should have a special prefix, such as "is/should/e.t.c"<br>Fix:                                                                                                                                                       | yes | no                                                                                                                                                                                                                         | Recursively update usages of this function in the projectAggressive fix - what if the new name will not be valid?                                                                                                                 |
```
Comment (vtchem) [inline info/available-rules.md:25]: The following sentence is not clear to me:
"Recursively update usages of this function in the projectAggressive fix - what if the new name will not be valid?"
Please, explain.

File: info/available-rules.md [hunk 13]
```diff
@@ -1,131 +1,131 @@
...
+| 2    | 2.1.1    | MISSING_KDOC_TOP_LEVEL                    | Check: warns at file level internal or public class or if the function has a missing KDoc<br>Fix: no                                                                                                                                                                 | no  | no                                                                                                                                                                                                                         | Support extension for setters/gettersSupport extension for method "override"                                                                                                                                                  |
```
Comment (vtchem) [inline info/available-rules.md]: "Support extension for setters/gettersSupport extension for method "override""
Please, explain.

File: info/available-rules.md [hunk 14]
```diff
@@ -1,131 +1,131 @@
...
+| 2    | 2.1.2    | KDOC_WITHOUT_PARAM_TAG                    | Check: warns if an accessible method has parameters and they are not documented in KDoc<br>Fix: If accessible method has no KDoc, the KDoc template is added                                                                                                        | yes | no                                                                                                                                                                                                                         | Should also make a separate fix, even if there is already some Kdoc in place                                                                                                                                                    |
```
Comment (vtchem) [inline info/available-rules.md]: "Should also make a separate fix, even if there is already some Kdoc in place" What does this sentence mean? Explanation is needed.

File: info/available-rules.md [hunk 15]
```diff
@@ -1,131 +1,131 @@
...
+| 2    | 2.1.3    | KDOC_NO_DEPRECATED_TAG                    | Check: warns if `@deprecated` is used in KDoc<br>Fix: adds `@Deprecated` annotation with a message, removes tag                                                                                                                                                | yes | no                                                                                                                                                                                                                         | The `replaceWith` field in the annotation can be filled too                                                                                                                                                                            |
```
Comment (vtchem) [inline info/available-rules.md:43]: "The `replaceWith` field in the annotation can be filled too." Is this statement correct?

File: info/available-rules.md [hunk 16]
```diff
@@ -1,131 +1,131 @@
...
+| 2    | 2.2.1    | HEADER_MISSING_IN_NON_SINGLE_CLASS_FILE   | Check: warns if a file with zero or >1 classes doesn't have a KDoc header                                                                                                                                                                                       | no  | no                                                                                                                                                                                                                         |                                                                                                                                                                                                                               |
```
Comment (vtchem) [inline info/available-rules.md:49]: warns if a file with zero or >1 classes
What does it mean? 
zero or >1 means any number of classes. Explanations needed.

File: info/available-rules.md [hunk 17]
```diff
@@ -1,131 +1,131 @@
...
+| 3    | 3.1.2    | FILE_WILDCARD_IMPORTS                     | Check: warns if the wildcard imports are used except allows.                                                                                                                                                                                                     | no  | allowedWildcards                                                                                                                                                                                                           | -                                                                                                                                                                                                                             |
```
Comment (vtchem) [inline info/available-rules.md:63]: "Check: warns if the wildcard imports are used except allows." Please explain.

File: info/available-rules.md [hunk 18]
```diff
@@ -1,131 +1,131 @@
...
+| 3    | 3.6.2    | COMPLEX_EXPRESSION                        | Check: warns if a long dot qualified expression is used in a condition or as an argument                                                                                                                                                                       | no  | no                                                                                                                                                                                                                         | -                                                                                                                                                                                                                             |
```
Comment (vtchem) [inline info/available-rules.md:77]: "Check: warns if a long dot qualified expression is used in a condition or as an argument"
What does "long dot qualified" mean?

File: info/available-rules.md [hunk 19]
```diff
@@ -1,131 +1,131 @@
...
+| 4    | 4.2.2    | TYPE_ALIAS                                | Check: if a type reference for property is longer than expected                                                                                                                                                                                                 | yes | typeReferenceLength                                                                                                                                                                                                        | -                                                                                                                                                                                                                             |
```
Comment (vtchem) [inline info/available-rules.md:100]: Duplicated line. Should we remove it?                                                                                                         |
| 4    | 4.2.2    | TYPE_ALIAS                                | Check: if the type reference for a property is longer than expected                                                                                                                                                                                                 | yes | typeReferenceLength

File: info/available-rules.md (hunk 17 above)
Comment (0x6675636b796f75676974687562) [reply info/available-rules.md:63]: This means that the so-called [wildcard imports](https://kotlinlang.org/spec/packages-and-imports.html), while accepted by the compiler, should be prohibited as per the _Kotlin Style Guide_, and any violations except for, well, exceptions, should be reported by _diKTat_ (as a part of `FILE_WILDCARD_IMPORTS` rule implementation).

If you look at `diktat-analysis.yml`, it adds a possiblility to configure `allowedWildcards`:

```yml
# Checks that there is no wildcard imports. Exception: allowedWildcards
- name: FILE_WILDCARD_IMPORTS
  enabled: true
  configuration:
    # Allowed wildcards for imports (e.g. "import org.cqfn.diktat.*, import org.jetbrains.kotlin.*")
    allowedWildcards: "kotlinx.serialization.*,org.cqfn.diktat.ruleset.utils.*"
```

File: info/available-rules.md (hunk 4 above)
Comment (0x6675636b796f75676974687562) [reply info/available-rules.md]: 2. This means that, technically, we can fix such issues automatically, but there's little sense in it. I believe you can remove thos whole sentence &mdash; it belongs to the bug tracker, **not** the user-visible documentation.

**Review by 0x6675636b796f75676974687562 — APPROVED** [review]

File: info/available-rules.md (hunk 2 above)
Comment (0x6675636b796f75676974687562) [reply info/available-rules.md]: >  (e. g. T12, T).

No, this example is misleading now.

`T12` is a good example, `T` is not (not followsed by numbers).

Also, please don't forget to surround code fragments with backticks, e.g.: `T12`.

File: info/available-rules.md (hunk 3 above)
Comment (0x6675636b796f75676974687562) [reply info/available-rules.md]: Do you mean the sentence per se? Or the formatting?

The formatting is mine, merely to emphasize the additions to the text. Being just an emphasis, the formatting should not propagate into the final document.

Speaking of the sentence -- yes, I do want this exact wording (possibly with your minor corrections) into the final proofread version.

File: info/available-rules.md (hunk 9 above)
Comment (0x6675636b796f75676974687562) [reply info/available-rules.md]: This is the last column, not intended to be read by users. So please just fix typos where appropriate, add a space after punctuation marks, and let's leave it as-is.

File: info/available-rules.md (hunk 10 above)
Comment (0x6675636b796f75676974687562) [reply info/available-rules.md]: This is about [package declarations](https://kotlinlang.org/docs/packages.html) in the Kotlin language.

This is correct:

```kotlin
package com.example.someModule
```

&mdash; while this is not:

```kotlin
package com.example.some_module
```

Neither is this:

```kotlin
package com.example.someModule_
```

File: info/available-rules.md (hunk 11 above)
Comment (0x6675636b796f75676974687562) [reply info/available-rules.md]: [`PascalCase`, `camelCase` and `snake_case` explained](https://khalilstemmler.com/blogs/camel-case-snake-case-pascal-case/).

Comment (0x6675636b796f75676974687562) [reply info/available-rules.md]: > Fix: fixing the case: if it is some fixed case (like snake/camel) - with word saving

This means, if the case is recognized to be a known one (Pascal, Camel or Snake), then we can detect word boundaries within the identifier name, and the fix is 100 percent accurate:

* `UPPER` -> `Upper`
* `lower` -> `Lower`
* `lowerCamel` -> `LowerCamel`
* `lower_snake` -> `LowerSnake`

> if not - will restore PascalCase as is

If the name of an identifier conforms to neither of the above 4 patterns, then a heuristic algorithm is applied:

* the 1st character, if lower-case, is capitalized: `c` -> `C`
* the character after an underscore is capitalized, and the underscore removed: `_c` -> `C`
* any upper-case character which follows another upper-case one is converted to the lower case: `AB` -> `Ab`

To summarize the above, an identifier named `paSCa_l` will be renbamed to `PaScaL`.

File: info/available-rules.md (hunk 12 above)
Comment (0x6675636b796f75676974687562) [reply info/available-rules.md]: > Recursively updates usages of this function in the project. Aggressive fix &mdash; what if the new function name will still be invalid?

File: info/available-rules.md (hunk 13 above)
Comment (0x6675636b796f75676974687562) [reply info/available-rules.md]: These are just to-do items, so let's not delve too much into this.

> Support extension for setters/getters. Support extension for overriding methods (those marked with the [`override`](https://kotlinlang.org/docs/inheritance.html#overriding-methods) keyword).

File: info/available-rules.md (hunk 14 above)
Comment (0x6675636b796f75676974687562) [reply info/available-rules.md]: Unclear to me either, let's leave as-is.

File: info/available-rules.md (hunk 15 above)
Comment (0x6675636b796f75676974687562) [reply info/available-rules.md]: We should consider not only annotating a class/method/property with `@Deprecated`, but also provide a meaningful `replaceWith` value, e.g.:

```kotlin
@Deprecated(replaceWith = "the code to be used as a replacement")
fun f() {
}
```

File: info/available-rules.md (hunk 18 above)
Comment (0x6675636b796f75676974687562) [reply info/available-rules.md]: This is a dot-qualified expression:

```kotlin
"".isEmpty
```

So is this:

```kotlin
object.functionCall()
```

Multiple function calls can be chained together:

```kotlin
object.foo().bar().baz()
```

If these are many, they should be split into multiple lines:

```kotlin
object.foo()
    .bar()
    .baz()
```

File: info/available-rules.md [hunk 20]
```diff
@@ -1,131 +1,131 @@
...
+| 4    | 4.1.3    | SAY_NO_TO_VAR                             | Check: warns if `var` modifier is used for a local variable (not in a class, not at file level). And this var is not used in accumulators                                                                                                              | no  | no                                                                                                                                                                                                                         | no                                                                                                                                                                                                                            |  several fixmes related to the search mechanism (VariablesSearch) and fixme for checking reassinment of this var |
```
[the remark this answers is not in the document]
Comment (0x6675636b796f75676974687562) [reply info/available-rules.md]: Please leave as-is.

File: info/available-rules.md (hunk 19 above)
Comment (0x6675636b796f75676974687562) [reply info/available-rules.md]: Yes, please.

Thanks for spotting this.

File: info/available-rules.md (hunk 2 above)
Comment (vtchem) [reply info/available-rules.md]: OK

File: info/available-rules.md (hunk 3 above)
Comment (vtchem) [reply info/available-rules.md]: OK. I changed it to read:

Check: warns if the name of a variable has a [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation) specific prefix (such as `mVariable` or `M_VARIABLE`), which is considered a bad code style (_Android_ is the only exception).

File: info/available-rules.md (hunk 4 above)
Comment (vtchem) [reply info/available-rules.md]: OK. Removed.

File: info/available-rules.md (hunk 9 above)
Comment (vtchem) [reply info/available-rules.md]: ok

File: info/available-rules.md (hunk 10 above)
Comment (vtchem) [reply info/available-rules.md]: I changed it to read:

Check: warns if the underscore is incorrectly used in package naming to split name parts.

File: info/available-rules.md (hunk 11 above)
Comment (vtchem) [reply info/available-rules.md]: OK. I slightly changed it:

Comment (vtchem) [reply info/available-rules.md]: If it is some fixed case (like Snake or Camel) - with word saving; if not - will restore PascalCase as is.

File: info/available-rules.md (hunk 12 above)
Comment (vtchem) [reply info/available-rules.md]: OK.

File: info/available-rules.md (hunk 13 above)
Comment (vtchem) [reply info/available-rules.md]: OK

File: info/available-rules.md (hunk 15 above)
Comment (vtchem) [reply info/available-rules.md]: Changed:

The `replaceWith` field in the annotation can also be filled with a meaningful value.

File: info/available-rules.md (hunk 17 above)
Comment (vtchem) [reply info/available-rules.md]: OK. I changed it as follows then:

Check: warns if the wildcard imports are used except allow**ed.**

Comment (vtchem) [reply info/available-rules.md]: Or:
Check: warns if the wildcard imports are used except the cases when they are allowed.

File: info/available-rules.md (hunk 18 above)
Comment (vtchem) [reply info/available-rules.md]: OK:

Comment (vtchem) [reply info/available-rules.md]: OK:

Check: warns if a long dot-qualified expression is used in a condition or as an argument.

File: info/available-rules.md (hunk 20 above)
[the remark this answers is not in the document]
Comment (vtchem) [reply info/available-rules.md]: OK

File: info/available-rules.md (hunk 19 above)
Comment (vtchem) [reply info/available-rules.md]: OK
